File: typescript/matrix/matrix.ts
export class Matrix {
  private _rows: number[][] = [];
  private _columns: number[][] = []

  constructor(matrix: string) {
    this.createRows(matrix);
    this.createColumns();
  }

  get rows(): number[][] {
    return this._rows;
  }

  get columns(): number[][] {
    return this._columns;
  }

  private createRows(matrix: string) {
    matrix.split('\n').forEach(row => {
      let columns = row.trim().split(' ').map(column => {
        return parseInt(column)
      });
      this._rows.push(columns);
    });
  }

  private createColumns() {
    for (let i = 0; i < this._rows[0].length; i++) {
      this._columns[i] = this._rows.map(row => { return row[i] });
    }
  }
}

File: ruby/matrix/matrix.rb
=begin
Write your code for the 'Matrix' exercise in this file. Make the tests in
`matrix_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/matrix` directory.
=end
class Matrix
  def initialize(matrix_string)
    @matrix_string = matrix_string
  end

  def row(number)
    rows[number - 1]
  end

  def column(number)
    columns[number - 1]
  end

  private

  def rows
    @rows ||= matrix_string.lines.map do |row|
      row.split.map(&:to_i)
    end
  end

  def columns
    @columns ||= rows.transpose
  end

  attr_reader :matrix_string
end